pkg/nimble/statconn/nimble_statconn.c

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright (C) 2020 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     pkg_nimble_statconn
+ * @{
+ *
+ * @file
+ * @brief       Statconn - static connection manager for NimBLE netif
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include "assert.h"
+#include "net/bluetil/addr.h"
+#include "net/bluetil/ad.h"
+#include "nimble_netif.h"
+#include "nimble_netif_conn.h"
+#include "nimble_statconn.h"
+
+#include "host/ble_hs.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define UNUSED          (0x00)
+#define PENDING         (0x80)
+#define CONNECTED       (0x00)
+#define ROLE_M          (0x01)
+#define ROLE_S          (0x02)
+
+typedef struct {
+    uint8_t addr[BLE_ADDR_LEN];     /**< peer addr, network byte order */
+    uint8_t state;                  /**< internal state */
+} slot_t;
+
+static const uint8_t _ad[2] = { BLE_GAP_AD_FLAGS, BLUETIL_AD_FLAGS_DEFAULT };
+
+static mutex_t _lock = MUTEX_INIT;
+static slot_t _slots[NIMBLE_NETIF_MAX_CONN];
+
+static struct ble_gap_adv_params _adv_params;
+static struct ble_gap_conn_params _conn_params;
+static uint32_t _conn_timeout;
+
+static nimble_netif_eventcb_t _eventcb = NULL;
+
+
+static slot_t *_get_addr(const uint8_t *addr)
+{
+    for (unsigned i = 0; i < ARRAY_SIZE(_slots); i++) {
+        if (memcmp(addr, _slots[i].addr, BLE_ADDR_LEN) == 0) {
+            return &_slots[i];
+        }
+    }
+    return NULL;
+}
+
+static slot_t *_get_state(uint8_t state)
+{
+    for (unsigned i = 0; i < ARRAY_SIZE(_slots); i++) {
+        if (_slots[i].state == state) {
+            return &_slots[i];
+        }
+    }
+    return NULL;
+}
+
+static void _activate(uint8_t role)
+{
+    mutex_lock(&_lock);
+    slot_t *slot = _get_state(role | PENDING);
+
+    if (slot && (role == ROLE_M)) {
+        ble_addr_t peer;
+        peer.type = BLE_ADDR_RANDOM;
+        bluetil_addr_swapped_cp(slot->addr, peer.val);
+        nimble_netif_connect(&peer, &_conn_params, _conn_timeout);
+    }
+    else if (slot && (role == ROLE_S)) {
+        nimble_netif_accept(_ad, sizeof(_ad), &_adv_params);
+    }
+    mutex_unlock(&_lock);
+}
+
+static void _update(const uint8_t *addr, uint8_t role, uint8_t state)
+{
+    mutex_lock(&_lock);
+    slot_t *slot = _get_addr(addr);
+    if (slot) {
+        slot->state = (role | state);
+    }
+    else {
+        DEBUG("[statconn] warning: state change on unknown peer address\n");
+    }
+    mutex_unlock(&_lock);
+
+    _activate(role);
+}
+
+static void _evt_dbg(const char *msg, int handle, const uint8_t *addr)
+{
+#if ENABLE_DEBUG
+    printf("[statconn] %s (%i|", msg, handle);
+    if (addr) {
+        bluetil_addr_print(addr);
+    }
+    else {
+        printf("n/a");
+    }
+    puts(")");
+#else
```

Comment:
If you use C conditionals here, you can remove skip the `(void)var`s and avoid ugly code:

```suggestion
    if (ENABLE_DEBUG) {
        printf("[statconn] %s (%i|", msg, handle);
        if (addr) {
            bluetil_addr_print(addr);
        }
        else {
            printf("n/a");
        }
        puts(")");
    }
```

The optimizer should take care of the rest.